File: server/app/schemas/cohorts.py
"""Cohorts V2 API schemas."""

from pydantic import BaseModel

from .base import (CohortMapping, DepartmentMapping, ProfileMapping,
                   SimulationMapping)
from .staff import StaffItem

# ============================================================================
# REQUEST SCHEMAS
# ============================================================================


class CohortsFilters(BaseModel):
    """Filters for cohorts list."""

    profileId: str


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================


class CohortItem(BaseModel):
    """Cohort item in list response."""

    cohort_id: str
    name: str  # Maps to cohorts.title
    description: str | None
    active: bool
    department_ids: list[str] | None  # None = cross-department (all departments)
    can_edit: bool
    can_delete: bool
    can_duplicate: bool
    can_leave: bool
    profile_ids: list[str]
    simulation_ids: list[str]
    num_members: int


class CohortsListResponse(BaseModel):
    """Response for cohorts list endpoint."""

    cohorts: list[CohortItem]
    profile_mapping: ProfileMapping
    simulation_mapping: SimulationMapping
    department_mapping: DepartmentMapping


# ============================================================================
# DETAIL SCHEMAS
# ============================================================================


class CohortDetailRequest(BaseModel):
    """Request for cohort detail."""

    cohortId: str
    profileId: str


class SimulationInCohort(BaseModel):
    """Simulation with cohort-specific statistics."""

    simulation_id: str
    name: str
    description: str
    time_limit: int | None
    active: bool

    # Cohort-specific statistics
    usage_count: int  # Number of attempts by cohort members
    success_rate: int  # Percentage (0-100) of graded attempts that passed
    last_used: str | None  # ISO timestamp or None
    can_remove: bool  # True if usage_count == 0


class CohortDetailResponse(BaseModel):
    """Response for cohort detail endpoint."""

    # Basic fields
    title: str  # cohorts.title
    description: str | None
    department_ids: list[str] | None  # None = cross-department (all departments)
    valid_department_ids: list[str]
    active: bool

    # Relationships
    simulation_ids: list[str]
    valid_simulation_ids: list[str]
    profile_ids: list[str]
    valid_profile_ids: list[str]

    # Full simulation objects with cohort-specific statistics
    simulations: list[SimulationInCohort]

    # Staff list (full ProfileListItem format)
    staff: list[StaffItem]

    # Top-level mappings
    simulation_mapping: SimulationMapping
    profile_mapping: ProfileMapping
    department_mapping: DepartmentMapping
    # Mappings for staff display (cohorts and departments)
    cohort_mapping: CohortMapping | None = None
    department_mapping_for_staff: DepartmentMapping | None = None


class CohortDetailDefaultRequest(BaseModel):
    """Request for default cohort detail."""

    profileId: str


# ============================================================================
# MUTATION SCHEMAS
# ============================================================================


class SimulationInRequest(BaseModel):
    """Simulation with active state for create/update requests."""

    simulation_id: str
    active: bool = True


class CreateCohortRequest(BaseModel):
    """Request to create cohort."""

    title: str
    description: str | None
    department_ids: list[str] | None  # None = cross-department (superadmin only)
    active: bool
    simulation_ids: list[str] | list[SimulationInRequest]  # Support both formats
    profile_ids: list[str]


class CreateCohortResponse(BaseModel):
    """Response from create cohort."""

    success: bool
    cohortId: str
    message: str


class UpdateCohortRequest(BaseModel):
    """Request to update cohort."""

    cohortId: str
    title: str
    description: str | None
    department_ids: list[str] | None  # None = cross-department (superadmin only)
    active: bool
    simulation_ids: list[str] | list[SimulationInRequest]  # Support both formats
    profile_ids: list[str]


class UpdateCohortResponse(BaseModel):
    """Response from update cohort."""

    success: bool
    message: str


class DuplicateCohortRequest(BaseModel):
    """Request to duplicate cohort."""

    cohortId: str


class DuplicateCohortResponse(BaseModel):
    """Response from duplicate cohort."""

    success: bool
    cohortId: str
    message: str


class DeleteCohortRequest(BaseModel):
    """Request to delete cohort."""

    cohortId: str


class DeleteCohortResponse(BaseModel):
    """Response from delete cohort."""

    success: bool
    message: str


class LeaveCohortRequest(BaseModel):
    """Request to leave cohort."""

    cohortId: str
    profileId: str


class LeaveCohortResponse(BaseModel):
    """Response from leave cohort."""

    success: bool
    message: str


class NewProfileForCohort(BaseModel):
    """New profile to create and add to cohort."""

    firstName: str
    lastName: str
    alias: str
    role: str


class AddProfilesToCohortRequest(BaseModel):
    """Request to add profiles to cohort (supports both existing and new)."""

    cohortId: str
    departmentIds: list[str]  # Needed for creating new profiles with dept relationships
    existingProfileIds: list[str] | None = None
    newProfiles: list[NewProfileForCohort] | None = None


class AddProfilesToCohortResponse(BaseModel):
    """Response from add profiles to cohort."""

    success: bool
    message: str


class RemoveProfilesFromCohortRequest(BaseModel):
    """Request to remove profiles from cohort."""

    cohortId: str
    profileIds: list[str]
    currentProfileId: str  # Current user's profile ID for permission validation


class RemoveProfilesFromCohortResponse(BaseModel):
    """Response from remove profiles from cohort."""

    success: bool
    message: str


# ============================================================================
# COHORT DETAIL WITH PROFILES (unified endpoint)
# ============================================================================


class CohortDetailWithProfilesRequest(BaseModel):
    """Request for cohort detail with available profiles."""

    cohortId: str
    departmentIds: list[str]
    currentProfileId: str


class CohortDetailWithProfilesResponse(BaseModel):
    """Response for cohort detail with available profiles."""

    # Cohort info
    cohort_id: str
    title: str
    description: str | None
    active: bool

    # Profile IDs already in cohort
    current_profile_ids: list[str]

    # Available profiles (filtered: instructional/ta, not in cohort, not default)
    available_profiles: list[StaffItem]

    # Mappings
    department_mapping: DepartmentMapping
    cohort_mapping: CohortMapping

File: client/lib/api/v2/schemas/cohorts.ts
/**
 * Cohorts V2 API Schemas
 * Schema definitions for cohorts v2 endpoints
 */

import { z } from "zod";
import {
  DepartmentMappingSchema,
  ProfileMappingSchema,
  SimulationMappingSchema,
} from "./base";
import { ProfileListItemSchema } from "./profile";

// ============================================================================
// REQUEST SCHEMAS
// ============================================================================

export const CohortsFiltersSchema = z.object({
  departmentIds: z.array(z.string()),
  profileId: z.string(),
});

export type CohortsFilters = z.infer<typeof CohortsFiltersSchema>;

// ============================================================================
// RESPONSE SCHEMAS
// ============================================================================

// Cohort item (normalized - IDs only)
export const CohortItemSchema = z.object({
  cohort_id: z.string(),
  name: z.string(), // Maps to cohorts.title
  description: z.string().nullable(),
  active: z.boolean(),
  default_cohort: z.boolean(),
  can_edit: z.boolean(),
  can_delete: z.boolean(),
  can_duplicate: z.boolean(),
  can_leave: z.boolean(),
  profile_ids: z.array(z.string()),
  simulation_ids: z.array(z.string()),
  num_members: z.number(),
});

export const CohortsListResponseSchema = z.object({
  cohorts: z.array(CohortItemSchema),
  profile_mapping: ProfileMappingSchema,
  simulation_mapping: SimulationMappingSchema,
});

export type CohortsListResponse = z.infer<typeof CohortsListResponseSchema>;
export type CohortItem = z.infer<typeof CohortItemSchema>;

// ============================================================================
// DETAIL SCHEMAS
// ============================================================================

export const CohortDetailRequestSchema = z.object({
  cohortId: z.string(),
  profileId: z.string(),
});

export type CohortDetailRequest = z.infer<typeof CohortDetailRequestSchema>;

export const CohortDetailResponseSchema = z.object({
  // Basic fields
  title: z.string(), // cohorts.title
  description: z.string().nullable(),
  department_id: z.string(),
  valid_department_ids: z.array(z.string()),
  active: z.boolean(),
  default_cohort: z.boolean(),

  // Relationships
  simulation_ids: z.array(z.string()),
  valid_simulation_ids: z.array(z.string()),
  profile_ids: z.array(z.string()),
  valid_profile_ids: z.array(z.string()),

  // Top-level mappings
  simulation_mapping: SimulationMappingSchema,
  profile_mapping: ProfileMappingSchema,
  department_mapping: DepartmentMappingSchema,
});

export type CohortDetailResponse = z.infer<typeof CohortDetailResponseSchema>;

// Default detail request
export const CohortDetailDefaultRequestSchema = z.object({
  profileId: z.string(),
});

export type CohortDetailDefaultRequest = z.infer<
  typeof CohortDetailDefaultRequestSchema
>;

// ============================================================================
// MUTATION SCHEMAS
// ============================================================================

// Create request
export const CreateCohortRequestSchema = z.object({
  title: z.string(),
  description: z.string().nullable(),
  department_id: z.string(),
  active: z.boolean(),
  default_cohort: z.boolean(),
  simulation_ids: z.array(z.string()),
  profile_ids: z.array(z.string()),
});

export type CreateCohortRequest = z.infer<typeof CreateCohortRequestSchema>;

export const CreateCohortResponseSchema = z.object({
  success: z.boolean(),
  cohortId: z.string(),
  message: z.string(),
});

export type CreateCohortResponse = z.infer<typeof CreateCohortResponseSchema>;

// Update request
export const UpdateCohortRequestSchema = z.object({
  cohortId: z.string(),
  title: z.string(),
  description: z.string().nullable(),
  department_id: z.string(),
  active: z.boolean(),
  default_cohort: z.boolean(),
  simulation_ids: z.array(z.string()),
  profile_ids: z.array(z.string()),
});

export type UpdateCohortRequest = z.infer<typeof UpdateCohortRequestSchema>;

export const UpdateCohortResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type UpdateCohortResponse = z.infer<typeof UpdateCohortResponseSchema>;

// Duplicate request
export const DuplicateCohortRequestSchema = z.object({
  cohortId: z.string(),
});

export type DuplicateCohortRequest = z.infer<
  typeof DuplicateCohortRequestSchema
>;

export const DuplicateCohortResponseSchema = z.object({
  success: z.boolean(),
  cohortId: z.string(),
  message: z.string(),
});

export type DuplicateCohortResponse = z.infer<
  typeof DuplicateCohortResponseSchema
>;

// Delete request
export const DeleteCohortRequestSchema = z.object({
  cohortId: z.string(),
});

export type DeleteCohortRequest = z.infer<typeof DeleteCohortRequestSchema>;

export const DeleteCohortResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type DeleteCohortResponse = z.infer<typeof DeleteCohortResponseSchema>;

// Leave cohort request
export const LeaveCohortRequestSchema = z.object({
  cohortId: z.string(),
  profileId: z.string(),
});

export type LeaveCohortRequest = z.infer<typeof LeaveCohortRequestSchema>;

export const LeaveCohortResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type LeaveCohortResponse = z.infer<typeof LeaveCohortResponseSchema>;

// New profile for cohort (to create and add)
export const NewProfileForCohortSchema = z.object({
  firstName: z.string(),
  lastName: z.string(),
  alias: z.string(),
  role: z.enum(["ta", "instructional"]),
});

export type NewProfileForCohort = z.infer<typeof NewProfileForCohortSchema>;

// Add profiles to cohort request (supports both existing and new)
export const AddProfilesToCohortRequestSchema = z.object({
  cohortId: z.string(),
  departmentIds: z.array(z.string()),
  existingProfileIds: z.array(z.string()).optional(),
  newProfiles: z.array(NewProfileForCohortSchema).optional(),
});

export type AddProfilesToCohortRequest = z.infer<
  typeof AddProfilesToCohortRequestSchema
>;

export const AddProfilesToCohortResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type AddProfilesToCohortResponse = z.infer<
  typeof AddProfilesToCohortResponseSchema
>;

// Remove profiles from cohort request
export const RemoveProfilesFromCohortRequestSchema = z.object({
  cohortId: z.string(),
  profileIds: z.array(z.string()),
});

export type RemoveProfilesFromCohortRequest = z.infer<
  typeof RemoveProfilesFromCohortRequestSchema
>;

export const RemoveProfilesFromCohortResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type RemoveProfilesFromCohortResponse = z.infer<
  typeof RemoveProfilesFromCohortResponseSchema
>;

// ============================================================================
// COHORT DETAIL WITH PROFILES (unified endpoint)
// ============================================================================

export const CohortDetailWithProfilesRequestSchema = z.object({
  cohortId: z.string(),
  departmentIds: z.array(z.string()),
  currentProfileId: z.string(),
});

export type CohortDetailWithProfilesRequest = z.infer<
  typeof CohortDetailWithProfilesRequestSchema
>;

export const CohortDetailWithProfilesResponseSchema = z.object({
  // Cohort info
  cohort_id: z.string(),
  title: z.string(),
  description: z.string().nullable(),
  active: z.boolean(),

  // Profile IDs already in cohort
  current_profile_ids: z.array(z.string()),

  // Available profiles (filtered: instructional/ta, not in cohort, not default)
  available_profiles: z.array(ProfileListItemSchema),

  // Mappings
  department_mapping: DepartmentMappingSchema,
  cohort_mapping: z.record(
    z.string(),
    z.object({
      name: z.string(),
      description: z.string(),
    })
  ),
});

export type CohortDetailWithProfilesResponse = z.infer<
  typeof CohortDetailWithProfilesResponseSchema
>;
